Use function `babbageMinUTxOValue` for faster minimum UTxO calculations.

jonathanknowles opened the pull request:
## Issue Number

ADP-2144

## Summary

This PR adjusts the `{compute,isBelow}MinimumCoinForUTxO` functions so that in the Babbage era only, we bypass the Cardano API function `calculateMinimumUTxO`, and instead directly use the ledger function `babbageMinUTxOValue`.

This leads to a significant reduction in latency when computing transaction fees:

**Before:**
```
    Latencies for 2 fixture wallets scenario
        postTransactionFee  - 3177.0 ms
    Latencies for 10 fixture wallets scenario
        postTransactionFee  - 3555.3 ms
```

**After:**
```
    Latencies for 2 fixture wallets scenario
        postTransactionFee  - 13.5 ms
    Latencies for 10 fixture wallets scenario
        postTransactionFee  - 15.6 ms
```

jonathanknowles committed:
Use function `babbageMinUTxOValue` for faster minimum UTxO calculations.

This commit adjusts the `{compute,isBelow}MinimumCoinForUTxO` functions
so that in the Babbage era only, we bypass the Cardano API function
`calculateMinimumUTxO`, and instead directly use the ledger function
`babbageMinUTxOValue`.

This leads to a significant reduction in latency when computing transaction
fees:

Before:
```
    Latencies for 2 fixture wallets scenario
        postTransactionFee  - 3177.0 ms
    Latencies for 10 fixture wallets scenario
        postTransactionFee  - 3555.3 ms
```

After:
```
    Latencies for 2 fixture wallets scenario
        postTransactionFee  - 13.5 ms
    Latencies for 10 fixture wallets scenario
        postTransactionFee  - 15.6 ms
```

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -174,3 +203,17 @@ unsafeCoinFromCardanoApiCalculateMinimumUTxOResult = \case
+    -> TokenBundle
+    -> Coin
+computeLedgerMinimumCoinForBabbage pp addr tokenBundle =
+    Coin $ fromIntegral @Integer @Natural result
```
This silently wraps negative values unlike [unsafeLovelaceToWalletCoin](https://github.com/input-output-hk/cardano-wallet/blob/c1f17e791ed9f69b752ef4359be1a1a940c5a3e7/lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs#L2146-L2155) for other eras. At least for consistency I think we should do it here as well.

jonathanknowles replied:
> This silently wraps negative values unlike [unsafeLovelaceToWalletCoin](https://github.com/input-output-hk/cardano-wallet/blob/c1f17e791ed9f69b752ef4359be1a1a940c5a3e7/lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs#L2146-L2155) for other eras. At least for consistency I think we should do it here as well.

Excellent point. I've fixed this in 529c7dee4b079b9333c319b14c91c69cab065376.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -66,11 +80,18 @@ computeMinimumCoinForUTxOShelleyBasedEra
+        _ ->
+            unsafeCoinFromCardanoApiCalculateMinimumUTxOResult $
+            Cardano.calculateMinimumUTxO era
+                (embedTokenMapWithinPaddedTxOut era addr tokenMap)
```
Don't think it matters and not sure if there is an obvious fix but FYI: perhaps slightly confusing that we're manually constructing `TokenBundle txOutMaxCoin` in one branch and calling `embedTokenMapWithinPaddedTxOut` in another branch.

jonathanknowles replied:
> Don't think it matters and not sure if there is an obvious fix but FYI: perhaps slightly confusing that we're manually constructing TokenBundle txOutMaxCoin in one branch and calling embedTokenMapWithinPaddedTxOut in another branch.

I agree it's not ideal!

I'd definitely like to tidy this up in a later PR, if possible.

Perhaps it won't be necessary to tidy though if we can work out why the slowdown happens when calling the Cardano API wrapper.

Anviking replied:
> Perhaps it won't be necessary to tidy though if we can work out why the slowdown happens when calling the Cardano API wrapper.

👍 Yeah, let's leave it. Besides going back to pure cardano-api it's also _possible_ that we realise we want to switch to ledger types also say after doing the balanceTx library switch.

Anviking reviewed (approved):
Awesome - lgtm! 🎉 

I'm kind of astonished how we'd be able to see this kind of extreme slowdown from the cardano-api wrapper... and only the babbage era 🤔 might laziness be involved somehow?

jonathanknowles committed:
Do not silently convert from Integer to Natural in `Shelley.MinimumUTxO`.

Instead, we use the pre-existing `Ledger.toWalletCoin` function, which
throws a run-time error if given a negative value.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3451#discussion_r950000165
